### PR TITLE
style-value-parser/feat: add parser support for more types of media queries

### PR DESCRIPTION
**Pull request by mellyeliu**
Lots of edge cases to fix:
- ~~support `only` keyword~~
- ~~calc parsing broken with certain parentheses~~
     - ~~implement `toString` in `Calc` parser class~~
- ~~duplicate not parsing broken~~
- ~~strange nested parentheses edge cases~~
- ~~refactor some width -> max-width logic~~
- ...to add as I test

Testing:
- so many tests! will refactor in a follow up PR
- added a number of invalid media query assertions
- i'm almost certain i'm missing some edge cases, but I think this covers 95+% of use cases and most of what we'll see practically. will continue to iterate

refs: [mdn media query docs](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_media_queries/Using_media_queries) 
[lightning css playground](https://lightningcss.dev/playground/index.html#%7B%22minify%22%3Afalse%2C%22customMedia%22%3Atrue%2C%22cssModules%22%3Afalse%2C%22analyzeDependencies%22%3Afalse%2C%22targets%22%3A%7B%22chrome%22%3A6225920%7D%2C%22include%22%3A0%2C%22exclude%22%3A0%2C%22source%22%3A%22%40custom-media%20--modern%20(color)%2C%20(hover)%3B%5Cn%5Cn%5Cn%40custom-media%20--%20(color)%2C%20(hover)%3B%5Cn%5Cn.foo%20%7B%5Cn%20%20background%3A%20yellow%3B%5Cn%5Cn%20%20-webkit-border-radius%3A%202px%3B%5Cn%20%20-moz-border-radius%3A%202px%3B%5Cn%20%20border-radius%3A%202px%3B%5Cn%5Cn%20%20-webkit-transition%3A%20background%20200ms%3B%5Cn%20%20-moz-transition%3A%20background%20200ms%3B%5Cn%20%20transition%3A%20background%20200ms%3B%5Cn%5Cn%20%20%26.bar%20%7B%5Cn%20%20%20%20color%3A%20green%3B%5Cn%20%20%7D%5Cn%7D%5Cn%5Cn%5Cn%5Cn%40media%20(min-width%3A%20500px)%20or%20(max-width%3A%20600px)%20and%20not%20(max-width%3A%2000px)%5Cn%20%7B%5Cn%20%20.a%20%7B%5Cn%20%20%20%20color%3A%20green%3B%5Cn%20%20%7D%5Cn%7D%5Cn%5Cn%40media%20print%20and%20((max-width%3A%20768px)%20or%20(orientation%3A%20portrait))%20%7B%5Cn%20%20.a%20%7B%5Cn%20%20%20%20color%3A%20green%3B%5Cn%20%20%7D%5Cn%7D%22%2C%22visitorEnabled%22%3Afalse%2C%22visitor%22%3A%22%7B%5Cn%20%20Color(color)%20%7B%5Cn%20%20%20%20if%20(color.type%20%3D%3D%3D%20'rgb')%20%7B%5Cn%20%20%20%20%20%20color.g%20%3D%200%3B%5Cn%20%20%20%20%20%20return%20color%3B%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%22%2C%22unusedSymbols%22%3A%5B%5D%2C%22version%22%3A%22local%22%7D) to double check syntax

### CODE REVIEW

File: packages/style-value-parser/src/at-queries-from-tokens/__tests__/media-query-test.js [hunk 1]
```diff
@@ -1874,7 +1892,7 @@ describe('Test CSS Type: @media queries', () => {
   });
 
   // nested not queries are broken
```
Comment (necolas) [inline packages/style-value-parser/src/at-queries-from-tokens/__tests__/media-query-test.js:1894]:
```suggestion
```

File: packages/style-value-parser/src/at-queries-from-tokens/__tests__/media-query-test.js [hunk 2]
```diff
@@ -1028,22 +1035,32 @@ describe('Test CSS Type: @media queries', () => {
...
-    expect(
-      MediaQuery.parser.parseToEnd('@media (width: calc(100% - 20px))'),
-    ).toMatchInlineSnapshot();
-  });
-
-  test.skip('@media (min-width: calc(300px + 5em))', () => {
-    expect(
-      MediaQuery.parser.parseToEnd('@media (min-width: calc(300px + 5em))'),
-    ).toMatchInlineSnapshot();
+  test('@media (min-width: calc(300px + 5em))', () => {
```
Comment (necolas) [inline packages/style-value-parser/src/at-queries-from-tokens/__tests__/media-query-test.js:1038]: Do we have any tests for when this MQ is malformed with no spaces around the operator within `calc`?

**Comment by mellyeliu** [comment]
yeah, the original intention was to build it reasonably close to the mdn spec so we have a parser approximating standard media query support, and we can limit what we want to include from there. for instance, I'm not sure if we actually want to support media query with calc (not sure how much use cases there are) but maybe we can parameterize it for flexibility moving forward

**Comment by nmn** [comment]
The parser is meant to help with many different things, so one of the things it might do is improve the ESLint plugin. There having a fully spec-compliant parser would mean the ability to offer auto-fixes.

File: packages/style-value-parser/src/at-queries-from-tokens/__tests__/media-query-test.js [hunk 3]
```diff
@@ -18,6 +18,7 @@ describe('Test CSS Type: @media queries', () => {
         "queries": {
           "key": "screen",
           "not": false,
+          "only": false,
```
Comment (nmn) [inline packages/style-value-parser/src/at-queries-from-tokens/__tests__/media-query-test.js:21]: I don't think it's valid to use `only` on any random query, and I think it can only exist on the `@media` level of the query.

File: packages/style-value-parser/src/at-queries-from-tokens/__tests__/media-query-test.js [hunk 4]
```diff
@@ -1028,22 +1035,32 @@ describe('Test CSS Type: @media queries', () => {
...
+    expect(MediaQuery.parser.parseToEnd(query)).toMatchInlineSnapshot(`
+      MediaQuery {
+        "queries": {
+          "key": "min-width",
+          "type": "pair",
+          "value": "calc(300px + 5em)",
+        },
+      }
+    `);
+    expect(MediaQuery.parser.parseToEnd(query).toString()).toEqual(query);
```
Comment (nmn) [inline packages/style-value-parser/src/at-queries-from-tokens/__tests__/media-query-test.js:1049]: I also think it's better to use snapshot tests instead of `toEqual(query)` as a pattern for these tests.

The parser, by design, will remove extra whitespace etc. It's job is to produce equivalent strings, not necessarily the same strings as the input. Instead of picking and choosing between the two patterns, it's better to use the safer pattern everywhere.

File: packages/style-value-parser/src/at-queries-from-tokens/media-query.js [hunk 5]
```diff
@@ -277,35 +295,103 @@ const mediaOrRulesParser: TokenParser<MediaOrRules | MediaQueryRule> =
...
+          ) {
+            const dim = queries.value as Length;
+            valueStr = `${dim.value}${dim.unit}`;
+          } else {
+            valueStr = candidate;
+          }
+        } else {
+          valueStr = '';
+        }
         return `(${queries.key}: ${valueStr})`;
```
Comment (nmn) [inline packages/style-value-parser/src/at-queries-from-tokens/media-query.js:390]: This long chain can be simplified with early returns.

```suggestion
        if (Array.isArray(queries.value)) {
          return `(${queries.key}: ${queries.value[0]} / ${queries.value[2]})`;
        }
        if (typeof queries.value === 'string') {
          return `(${queries.key}: ${queries.value})`;
        } 
        if (
          typeof queries.value === 'object' &&
          queries.value !== null &&
          typeof queries.value.toString === 'function'
        ) {
          const candidate = queries.value.toString();
          if (
            candidate === '[object Object]' &&
            typeof queries.value !== 'string' &&
            !Array.isArray(queries.value)
          ) {
            const dim = queries.value as Length;
            return `(${queries.key}: ${dim.value}${dim.unit})`;
          }
          return `(${queries.key}: ${candidate})`;
        } 
        return `(${queries.key}: )`;
```

This reveals a problem with the last `else` case where `(${queries.key}: )` doesn't seem like a valid thing to return.

File: packages/style-value-parser/src/at-queries-from-tokens/media-query.js [hunk 6]
```diff
@@ -277,35 +295,103 @@ const mediaOrRulesParser: TokenParser<MediaOrRules | MediaQueryRule> =
...
+        } else if (typeof queries.value === 'string') {
+          valueStr = queries.value;
+        } else if (
+          typeof queries.value === 'object' &&
+          queries.value !== null &&
+          typeof queries.value.toString === 'function'
+        ) {
+          const candidate = queries.value.toString();
+          if (
+            candidate === '[object Object]' &&
```
Comment (nmn) [inline packages/style-value-parser/src/at-queries-from-tokens/media-query.js:378]: `'[object Object]'` is also a bit sketchy.

File: packages/style-value-parser/src/at-queries-from-tokens/media-query.js [hunk 7]
```diff
@@ -277,35 +295,103 @@ const mediaOrRulesParser: TokenParser<MediaOrRules | MediaQueryRule> =
...
+    () =>
+      mediaOrRulesParser.surroundedBy(
+        TokenParser.tokens.OpenParen,
+        TokenParser.tokens.CloseParen,
+      ),
+    () =>
+      mediaAndRulesParser.surroundedBy(
+        TokenParser.tokens.OpenParen,
+        TokenParser.tokens.CloseParen,
+      ),
```
Comment (nmn) [inline packages/style-value-parser/src/at-queries-from-tokens/media-query.js:334]: Pretty sure that you also need to account for optional whitespace here.

**Review by nmn — CHANGES_REQUESTED** [review]
Some issues need fixing.

File: packages/style-value-parser/src/at-queries-from-tokens/__tests__/media-query-test.js (hunk 3 above)
Comment (mellyeliu) [reply packages/style-value-parser/src/at-queries-from-tokens/__tests__/media-query-test.js:21]: that's how it's currently implemented, `only` is only allowed at the top level of a query, not within params or within a nested clause. i'll add some tests to cover it

File: packages/style-value-parser/src/at-queries-from-tokens/__tests__/media-query-test.js (hunk 4 above)
Comment (mellyeliu) [reply packages/style-value-parser/src/at-queries-from-tokens/__tests__/media-query-test.js:1049]: makes sense, I agree. I'm going to do a larger refactor to organize these tests in the next PR, and will rewrite how we test toString as well

File: packages/style-value-parser/src/at-queries-from-tokens/media-query.js (hunk 5 above)
Comment (mellyeliu) [reply packages/style-value-parser/src/at-queries-from-tokens/media-query.js]: added a throw in that case, let's support this set of values first

**Review by nmn — APPROVED** [review]
I don't see any issues for now. We can iterate more if any issues arise.
